SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp

```diff
@@ -620,35 +706,227 @@ void ASpatialDebugger::DrawDebug(UCanvas* Canvas, APlayerController* /* Controll
-				}
+				DrawTag(Canvas, ScreenLocation, EntityId, Actor->GetName(), true /*bCentre*/);
+			}
+		}
+	}
+}
+
+void ASpatialDebugger::SelectActorsToTag(UCanvas* Canvas)
+{
+	if (LocalPlayerController.IsValid())
+	{
+		FVector2D NewMousePosition;
 
-				FVector2D ScreenLocation = FVector2D::ZeroVector;
-				if (LocalPlayerController.IsValid())
+		if (LocalPlayerController->GetMousePosition(NewMousePosition.X, NewMousePosition.Y))
+		{
+			if (CrosshairTexture != nullptr)
+			{
+				// Display a crosshair icon for the mouse cursor
+				// Offset by half of the texture's dimensions so that the center of the texture aligns with the center of the Canvas.
+				FVector2D CrossHairDrawPosition(NewMousePosition.X - (CrosshairTexture->GetSurfaceWidth() * 0.5f),
+												NewMousePosition.Y - (CrosshairTexture->GetSurfaceHeight() * 0.5f));
+
+				// Draw the crosshair at the mouse position.
+				FCanvasTileItem TileItem(CrossHairDrawPosition, CrosshairTexture->Resource, FLinearColor::White);
+				TileItem.BlendMode = SE_BLEND_Translucent;
+				Canvas->DrawItem(TileItem);
+			}
+
+			TWeakObjectPtr<AActor> NewHoverActor = GetActorAtPosition(NewMousePosition);
+			HighlightActorUnderCursor(NewHoverActor);
+		}
+
+		// Draw tags above selected actors
+		for (TWeakObjectPtr<AActor> SelectedActor : SelectedActors)
+		{
+			if (SelectedActor.IsValid())
+			{
+				if (const Worker_EntityId_Key* HitEntityId = EntityActorMapping.FindKey(SelectedActor))
 				{
-					SCOPE_CYCLE_COUNTER(STAT_Projection);
-					UGameplayStatics::ProjectWorldToScreen(LocalPlayerController.Get(), ActorLocation + WorldSpaceActorTagOffset,
-														   ScreenLocation, false);
+					FVector PlayerLocation = GetLocalPawnLocation();
+
+					FVector2D ScreenLocation = ProjectActorToScreen(SelectedActor, PlayerLocation);
+					if (!ScreenLocation.IsZero())
+					{
+						DrawTag(Canvas, ScreenLocation, *HitEntityId, SelectedActor->GetName(), true /*bCentre*/);
+					}
 				}
+			}
+		}
+	}
+}
 
-				if (ScreenLocation.IsZero())
+void ASpatialDebugger::HighlightActorUnderCursor(TWeakObjectPtr<AActor>& NewHoverActor)
+{
+	if (!bShowHighlight)
+	{
+		return;
+	}
+
+	if (!NewHoverActor.IsValid())
+	{
+		// No actor under the cursor so revert hover materials on previous actor
+		RevertHoverMaterials();
+	}
+	else if (NewHoverActor != HoverActor)
+	{
+		// New actor under the cursor
+
+		// Revert hover materials on previous actor
+		RevertHoverMaterials();
+
+		// Set hover materials on new actor
+		TArray<UActorComponent*> ActorComponents;
+		NewHoverActor->GetComponents(UMeshComponent::StaticClass(), ActorComponents, true);
+		for (UActorComponent* NewActorComponent : ActorComponents)
+		{
+			// Store previous components
+			TWeakObjectPtr<UMeshComponent> MeshComponent(Cast<UMeshComponent>(NewActorComponent));
+			TWeakObjectPtr<UMaterialInterface> MeshMaterial = MeshComponent->GetMaterial(0);
+			if (MeshComponent.IsValid() && MeshMaterial.IsValid() && WireFrameMaterial != nullptr)
```

Comment:
WireFrameMaterial != nullptr can be checked at the top of the function

Reply:
Sure, moved the check to the top of the function.